### ISSUE TITLE
`or` and `and` in await statements

### ISSUE THREAD

**Issue by mageowl** [godotengine/godot-proposals#6243]
### Describe the project you are working on

A mobile auto-battler about ducks

### Describe the problem or limitation you are having in your project

I cannot wait for the confirm *or* deny button to be pressed. You are able to wait for one to be pressed, but not the other.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allow people to use the logic operators `or` and `and` on signals in await statements (`or` is wait for one to finish, `and` is wait for all to finish.)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```gdscript
await button_1.pressed or button_2.pressed
```
(Wait for one to finish)

---
```gdscript
await button_1.pressed and button_2.pressed
```
(Wait for all of them to finish)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

You could create a new signal that is triggered when one button is pressed.

```gdscript
signal one_of_them_pressed

func _on_any_button_pressed():
   one_of_them_pressed.emit()
```

### Is there a reason why this should be core and not an add-on in the asset library?

Add-ons are unable to modify GDScript.

**Comment by TheDuriel** [comment]
On one hand, this can always be worked around with through additional code. Like a third signal. But also it invites some interesting possibilities.

One problem though is: How does this behave with return values? I suppose in an "and" case, we generate a 2D array like with call_group?

**Comment by Calinou** [comment]
Using `or`/`and` for this purpose might feel strange. What about doing something similar to JavaScript's `Promise.all()`/`Promise.any()`, where you can create a new signal that is emitted when both (or any) are satisfied?

**Comment by TheDuriel** [comment]
I think at that point, implementation becomes a bit pointless. Since you can easily implement a handler for that yourself.

```
class_name Promise
extends Reference

signal completed(args) # array

enum MODE {ANY, ALL}
var mode: int = MODE.ANY

var _signals: Dictionary = {} # signal: {has_emitted: bool, data: any}
var _completed: bool = false


func _init(signals: Array, mode: int) -> void:
	for s in signals:
		s.connect(self, _on_signal, [s])


func _on_signal(signal: Callable?, arg1: void, arg2: void, etc., placeholders) -> void:
	_signals[signal].has_emitted = true
	_signals[signal].data = [arg1, and so on] # There's ways to clean this up.
	
	_check_completion()


func _check_completion():
	if _completed:
		return

	if mode == MODE.ANY:
		_check_any_completion()
	elif mode == MODE.:
		_check_all_completion()


func _check_any_completion() -> void:
	for signal in _signals:
		if _signals[signal].has_emitted:
			completed.emit(_signals[signal].data)
			return


func _check_all_completion() -> void:
	var return_data: Array = []
	
	for signal in _signals:
		if not _signals[signal].has_emitted:
			return
		else:
		return_data.append(_signals[signal].data)
	
```

**Comment by mageowl** [comment]
> On one hand, this can always be worked around with through additional code. Like a third signal. But also it invites some interesting possibilities.
> 
> One problem though is: How does this behave with return values? I suppose in an "and" case, we generate a 2D array like with call_group?

It could return an array with the different return values.

```gdscript
func abc()
  var arr = await toggle.input and btn.input
  # arr = [event1, event2]

```

**Comment by dalexeev** [comment]
As mentioned above, this can be worked around quite easily. You can create the class/method you need and pass signals/functions into it (also note the `Callable.bind` method). You can find examples in related proposals:

* #2518
* #2622
* #3469
* #3964
* #4889
* #5510

**Comment by JoNax97** [comment]
> I think this is bad syntax because it makes the behavior of `and` and `or` keywords context dependent. Normally, `signal1 or signal2` results in an error because `Signal` is an invalid type for the `or` operator.

Not necessarily. You can make those operators always return a new callable that groups the operands (basically what @TheDuriel described, but automatically)

**Comment by tebz** [comment]
await_any( [a,b,c] )
I'd like it to return the index of the signal…

**Comment by SlugFiller** [comment]
> Using `or`/`and` for this purpose might feel strange. What about doing something similar to JavaScript's `Promise.all()`/`Promise.any()`, where you can create a new signal that is emitted when both (or any) are satisfied?

I should probably mention that JS's `Promise.race`/`Promise.any` leaks memory. Because JS accepts any `thenable` as a promise, if there's a race between a promise that fulfills, to one that never settles, the latter retains the handler long after the race has settled, even though the handler is set up to simply ignore a settlement at that point.

Presumably, this is not an issue in Godot due to `Signal.disconnect`. But, just as an example, @TheDuriel 's implementation above is leaking memory. This is something that needs to be taken into consideration, since `await` works as a one-shot. If not careful, it's easy to create a highly unwanted reference loop between emitter and waiter.

The sheer foot-gun potential alone of a working-but-leaky implementation justifies an official implementation.

**Comment by TheDuriel** [comment]
No it doesn't? And indeed your JS concern is wholly irrelevant to Godot as a whole.

**Comment by TheDuriel** [comment]
Disconnection occurs automatically if either object is freed. The promise object extends Reference, which is automatically freed if there are no more references to it. **Signals do not create references.**

This cleans itself up completely and without your own intervention.

You've additionally completely missed the nature of sudo code...

**Comment by HaarMegiddo** [comment]
Any updates on this? 
Don't really have an opinion on the syntax but this really feels like it should be a built in feature instead of making custom implementations With additional signals for this behaviour.

**Comment by HaarMegiddo** [comment]
Also, one more important thing should be built-in into this. Whenever `any` or `all` call returns, the user should also be able to get the arguments the signal was called with. And in case of `any` we should also be able to get the exact signal that was returned.

**Comment by epicanemone28** [comment]
suprised this hasn't been made a part of godot, given parallel timed sequences are a big part of why asynchronous stuff is so helpful. i definitely agree with the idea of riffing on javascript's promise.any and promise.all constructors, it's a very nice way of working with them.

**Comment by Oblongmana** [comment]
There might be some meta lessons from Javascript Promises as well which I'm vaguely familiar with from my former life as a web dev - there were a lot of different promise libraries with all sorts of different idiosyncracies, bugs, etc. 

As the use of Promises proliferated incredibly widely (as it's a very useful pattern!), it was eventually rolled into the ECMAScript standard itself. This brought great efficiency improvements, standardisation, and I'd argue improved web stability in the aggregate as it reduced the amount of "roll-your-own" required - to the extent that two of the former leading Promise libraries now plead that users use native JS Promises! See Bluebird: https://github.com/petkaantonov/bluebird/, and Q (recently archived too!): https://github.com/kriskowal/q)
